// express-validator/index.js
function getValue(req, location, field) {
  const source = location === 'params' ? req.params : location === 'query' ? req.query : req.body;
  return source ? source[field] : undefined;
}

function isEmptyValue(value) {
  return value === undefined || value === null || (typeof value === 'string' && value.trim() === '');
}

function createValidationChain(location, field) {
  const chain = function validationMiddleware(req, res, next) {
    const source = location === 'params' ? req.params : location === 'query' ? req.query : req.body;
    if (!source) {
      return next();
    }

    if (!req._expressValidatorErrors) {
      req._expressValidatorErrors = [];
    }

    const value = source[field];
    if (chain.optionalFlag && (value === undefined || value === null || value === '')) {
      return next();
    }

    let currentValue = value;
    for (const check of chain.checks) {
      try {
        currentValue = check(currentValue);
      } catch (error) {
        req._expressValidatorErrors.push({
          path: field,
          msg: chain.message || error.message,
          location,
        });
        break;
      }
    }

    if (source && Object.prototype.hasOwnProperty.call(source, field) && currentValue !== undefined) {
      source[field] = currentValue;
    }

    return next();
  };

  chain.location = location;
  chain.field = field;
  chain.checks = [];
  chain.optionalFlag = false;
  chain.message = null;

  chain.optional = function optional() {
    chain.optionalFlag = true;
    return chain;
  };

  chain.trim = function trim() {
    chain.checks.push((value) => (typeof value === 'string' ? value.trim() : value));
    return chain;
  };

  chain.notEmpty = function notEmpty() {
    chain.checks.push((value) => {
      if (isEmptyValue(value)) {
        throw new Error(chain.message || 'Value is required');
      }
      return value;
    });
    return chain;
  };

  chain.isLength = function isLength(options) {
    chain.checks.push((value) => {
      if (isEmptyValue(value)) {
        return value;
      }
      const length = value.length;
      const min = options && options.min !== undefined ? options.min : null;
      const max = options && options.max !== undefined ? options.max : null;
      if ((min !== null && length < min) || (max !== null && length > max)) {
        throw new Error(chain.message || 'Invalid length');
      }
      return value;
    });
    return chain;
  };

  chain.isInt = function isInt(options) {
    chain.checks.push((value) => {
      if (isEmptyValue(value)) {
        return value;
      }
      const parsed = Number(value);
      const min = options && options.min !== undefined ? options.min : null;
      const max = options && options.max !== undefined ? options.max : null;
      if (!Number.isInteger(parsed) || (min !== null && parsed < min) || (max !== null && parsed > max)) {
        throw new Error(chain.message || 'Must be an integer');
      }
      return value;
    });
    return chain;
  };

  chain.isFloat = function isFloat(options) {
    chain.checks.push((value) => {
      if (isEmptyValue(value)) {
        return value;
      }
      const parsed = Number(value);
      const min = options && options.min !== undefined ? options.min : null;
      if (Number.isNaN(parsed) || (min !== null && parsed < min)) {
        throw new Error(chain.message || 'Must be a number');
      }
      return value;
    });
    return chain;
  };

  chain.isMongoId = function isMongoId() {
    chain.checks.push((value) => {
      if (isEmptyValue(value)) {
        return value;
      }
      if (!/^[0-9a-fA-F]{24}$/.test(String(value))) {
        throw new Error(chain.message || 'Invalid id');
      }
      return value;
    });
    return chain;
  };

  chain.isIn = function isIn(values) {
    chain.checks.push((value) => {
      if (isEmptyValue(value)) {
        return value;
      }
      if (!values.includes(value)) {
        throw new Error(chain.message || 'Invalid value');
      }
      return value;
    });
    return chain;
  };

  chain.withMessage = function withMessage(message) {
    chain.message = message;
    return chain;
  };

  return chain;
}

function body(field) {
  return createValidationChain('body', field);
}

function param(field) {
  return createValidationChain('params', field);
}

function query(field) {
  return createValidationChain('query', field);
}

function validationResult(req) {
  if (!req._expressValidatorErrors) {
    req._expressValidatorErrors = [];
  }

  return {
    isEmpty() {
      return req._expressValidatorErrors.length === 0;
    },
    array() {
      return req._expressValidatorErrors;
    },
  };
}

module.exports = {
  body,
  param,
  query,
  validationResult,
};
